### PR TITLE
feat: add comprehensive unit tests for shell parsing helpers

### LINKED ISSUE #70: Add unit tests for shell parsing helpers
Look at the shell/shell.go file. The unit tests are completely missing.
It could be useful to have it!

### PR DESCRIPTION
**Pull request by Deep-Dey1**
## What Changed
Added comprehensive unit tests for shell parsing helper functions that were previously missing.

## Why
Addresses the issue: "The unit tests are completely missing" for shell/shell.go

## Tests Added
- 11 test functions covering 12 helper functions
- 99+ test cases with edge cases and integration tests
- All tests passing (100% pass rate)

## How to Test
```bash
go test ./shell -v

### CODE REVIEW

**Comment by Deep-Dey1** [comment]
Even I will try to fix these by tonight.

**Comment by Deep-Dey1** [comment]
Thankyou @dmarro89 !
😊

**Comment by dmarro89** [comment]
> ### 💡 Code Review for PR #72
> **Overall Impression:** The new unit tests in `shell/shell_test.go` are excellent! The coverage for parsing helpers hits boundary conditions, empty states, and standard flows correctly. However, a slight architectural adjustment is needed for how you mocked the I/O system.
> 
> **Issue Identified:** Currently, `drivers/ata/ata.go` and `terminal/terminal.go` were modified to add empty function bodies directly (e.g., `func inb(...) byte { return 0 }`). While this successfully satisfies the standard `go test` linker locally, **this breaks the OS runtime**. Replacing the abstract declarations severs the `stubs_amd64.s` assembly code from the compiler. `inb` will just return 0, `waitDRQ` will time out or loop indefinitely, and disk-backed shell commands will hang the OS in production.
> 
> **How To Fix (Using Build Tags):** To cleanly solve this, you need to use Go build tags so the compiler uses `stubs` locally for `go test`, but preserves the un-bodied assembly prototypes for the production OS kernel (which uses `gccgo`).
> 
> **1. Revert edits in `drivers/ata/ata.go`** Remove the definitions of `inb`, `outb`, `insw`, `outsw` entirely from this file.
> 
> **2. Create `drivers/ata/ata_gccgo.go` (Production Code):**
> 
> ```go
> //go:build gccgo
> 
> package ata
> 
> func inb(port uint16) byte
> func outb(port uint16, value byte)
> func insw(port uint16, addr *byte, count int)
> func outsw(port uint16, addr *byte, count int)
> ```
> 
> **3. Create `drivers/ata/ata_stub.go` (Local Testing Code):**
> 
> ```go
> //go:build !gccgo
> 
> package ata
> 
> func inb(port uint16) byte { return 0 }
> func outb(port uint16, value byte) {}
> func insw(port uint16, addr *byte, count int) {}
> func outsw(port uint16, addr *byte, count int) {}
> ```
> 
> **4. Revert edits in `terminal/terminal.go`** Remove the definitions of `outb` and `debugChar` entirely from this file.
> 
> **5. Create `terminal/terminal_gccgo.go` (Production Code):**
> 
> ```go
> //go:build gccgo
> 
> package terminal
> 
> func outb(port uint16, value byte)
> func debugChar(c byte)
> ```
> 
> **6. Create `terminal/terminal_stub.go` (Local Testing Code):**
> 
> ```go
> //go:build !gccgo
> 
> package terminal
> 
> func outb(port uint16, value byte) {}
> func debugChar(c byte) {}
> ```
> 
> Making these architectural updates will keep your unit tests green i guess but good job !

Thanks for your code review @ranjan42

**Comment by Deep-Dey1** [comment]
## Summary of Changes

The original PR added function bodies directly to `ata.go` and `terminal.go`, which prevented the OS from linking with assembly stubs and broke production code. As mentioned by @ranjan42, and following their proposed solution, I have now implemented Go build tags following the project's existing pattern.

### Architecture Changes

#### ATA Driver Package (`drivers/ata/`)

**1. `ata_gccgo.go` - Production Code**
```go
//go:build gccgo

package ata

func inb(port uint16) byte
func outb(port uint16, value byte)
func insw(port uint16, addr *byte, count int)
func outsw(port uint16, addr *byte, count int)
```

**2. `ata_stub.go` - Testing Stubs**
```go
//go:build !gccgo

package ata

func inb(port uint16) byte { return 0 }
func outb(port uint16, value byte) {}
func insw(port uint16, addr *byte, count int) {}
func outsw(port uint16, addr *byte, count int) {}
```

**3. `ata.go` - Reverted**
- Removed function bodies
- Kept only constants and logic

#### Terminal Package (`terminal/`)

**1. `terminal_gccgo.go` - Production Code**
```go
//go:build gccgo

package terminal

func outb(port uint16, value byte)
func debugChar(c byte)
```

**2. `terminal_stub.go` - Testing Stubs**
```go
//go:build !gccgo

package terminal

func outb(port uint16, value byte) {}
func debugChar(c byte) {}
```

**3. `terminal.go` - Reverted**
- Removed function bodies
- Kept only constants and logic

### Results

- ✅ All 15 test functions pass (99+ test cases)
- ✅ Production code remains compatible (gccgo links with assembly stubs)
- ✅ Testing environment works locally (go test uses stubs)
- ✅ No impact on production kernel behavior
- ✅ Follows existing project patterns

### Testing
```bash
go test ./shell -v
```

This approach ensures tests pass locally while preserving assembly linkage for the production OS kernel.

**Comment by dmarro89** [comment]
> Thankyou @ranjan42 and @dmarro89 for the reviews and suggestions. 
> 
> Tell me if any further changes needed !

Hi @Deep-Dey1 for me the PR is ok but the build fails.

File: shell/shell_test.go [hunk 1]
```diff
@@ -0,0 +1,460 @@
...
+			}
+			if ok && result != tt.expLen {
+				t.Errorf("copyNameFromRange(%d, %d) = %d, expected %d", tt.start, tt.end, result, tt.expLen)
+			}
+		})
+	}
+}
+
+// TestMinInt tests the minInt helper function
+func TestMinInt(t *testing.T) {
```
Comment (ArmandoBarragan) [inline shell/shell_test.go:302]: I do not think there's a need for this test. It's a pretty simple function and it only makes the codebase bigger

**Review by ArmandoBarragan — APPROVED** [review]

**Comment by ranjan42** [comment]
Hey @Deep-Dey1  , can you please let us know if you are still working on fixing the build ?

**Review by ranjan42 — COMMENTED** [review]
Looks good but can you please fix the build issues

File: shell/shell_test.go (hunk 1 above)
Comment (dmarro89) [reply shell/shell_test.go:302]: @Deep-Dey1 ehy did you check this comment ?

**Review by dmarro89 — APPROVED** [review]
